[CPU] Add torch.trace for complex tensors

anjali411 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50380 [CPU] Add torch.trace for complex tensors**

Differential Revision: [D25949361](https://our.internmc.facebook.com/intern/diff/D25949361)

anjali411 committed:
[CPU] Add torch.trace for complex tensors

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -504,30 +504,48 @@ Tensor trace_cpu(const Tensor& self) {
-    // all integer types get promoted to kLong
-    if (result.scalar_type() == at::kLong) {
+      // all integer types get promoted to kLong
       *result.data_ptr<int64_t>() = sum;
```
separated dispatch because this cast causes a compile error for complex otherwise

ezyang replied:
This seems like a good opportunity to use a `if_constexpr` to fix the problem. Test for `std::is_integral<scalar_t>` . Check `if_constexpr` docs for how to use it.

anjali411 replied:
still gives a compile error

ezyang commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -504,30 +504,48 @@ Tensor trace_cpu(const Tensor& self) {
-    using accscalar_t = at::acc_type<scalar_t, false>;
-    accscalar_t sum = 0;
-    const auto* t_data = self.data_ptr<scalar_t>();
+  if (result.scalar_type() == at::kLong) {
```
Hm, is this test right? What about other integral types besides Long?

mruberry replied:
Surprisingly this is kinda the right test.

```
if (dtype == at::kLong) {
  ...
}
```

might be clearer. Add a comment here explaining that all integer types (including bool) will return kLong from `get_dtype()`, and that this branch handles integer types.

mruberry replied:
Why is the new branch needed, however? It seems like, as in the previous kernel, the only different is at the assignment to result at the end?

anjali411 replied:
yeah it is a bit confusing with no comment, so I added one for clarity.

anjali411 replied:
@mruberry this is why https://github.com/pytorch/pytorch/pull/50380/files#r555248150

mruberry commented on test/test_torch.py:
```diff
@@ -6843,7 +6843,7 @@ def inner(self, device, dtype):
     ('topk', 'dim_desc_sort', _small_3d_unique, lambda t, d: [2, 1, True, True],
         1e-5, 1e-5, 1e-5, torch.testing.get_all_dtypes(include_complex=False, include_bool=False), _cpu_types, False),
-    ('trace', '', _medium_2d, lambda t, d: [], 1e-3, 1e-5, 1e-5, _types, _cpu_types, False),
+    ('trace', '', _medium_2d, lambda t, d: [], 1e-3, 1e-5, 1e-5, _types + _complex_types, _cpu_types, False),
```
Implement an OpInfo for trace and delete this entry, update `test_trace` in test_shape_ops to include complex types, and delete its method_test entry:

https://github.com/pytorch/pytorch/blob/057be231688f712212688530f0105ba143e9a1bf/test/test_shape_ops.py#L228

https://github.com/pytorch/pytorch/blob/057be231688f712212688530f0105ba143e9a1bf/torch/testing/_internal/common_methods_invocations.py#L2221

anjali411 replied:
done!

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"




[ghstack-poisoned]

anjali411 commented:
@mruberry I updated the PR based on your comments. could you take a look?

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"




[ghstack-poisoned]

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"




[ghstack-poisoned]

mruberry commented:
Looks like CUDA supports bool and bfloat16, so those need to be added to the OpInfo's CUDA dtypes.

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"


Differential Revision: [D25949361](https://our.internmc.facebook.com/intern/diff/D25949361)

[ghstack-poisoned]

ezyang reviewed (approved):
Approving to unblock, but please give the `if_constexpr` trick a try, I think you can avoid the duplication with it.

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"


Differential Revision: [D25949361](https://our.internmc.facebook.com/intern/diff/D25949361)

[ghstack-poisoned]

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"


Differential Revision: [D25949361](https://our.internmc.facebook.com/intern/diff/D25949361)

[ghstack-poisoned]

anjali411 commented on aten/src/ATen/native/ReduceOps.cpp:
```diff
@@ -521,12 +524,11 @@ Tensor trace_cpu(const Tensor& self) {
+    c10::guts::if_constexpr<std::is_integral<accscalar_t>::value>(
+      // all integer types get promoted to kLong
+      [&] (auto _) { *result.data_ptr<int64_t>() = sum; },  // then-case, invalid for non-integral types
+      [&] (auto _) { *result.data_ptr<scalar_t>() = sum; }  // else-case, invalid for integral types
```
cc. @ezyang the build still fails

smessmer replied:
The compiler is eagerly type-checking the else-case even if the condition is true (and the other way round). You need to prevent the compiler from doing so, so that it can only type-check the branch that's actually called. You can do this by wrapping things into the `_` you get passed as an argument. See https://github.com/pytorch/pytorch/blob/master/c10/util/C%2B%2B17.h#L257

anjali411 committed:
Update on "[CPU] Add torch.trace for complex tensors"


Differential Revision: [D25949361](https://our.internmc.facebook.com/intern/diff/D25949361)

[ghstack-poisoned]
